cmd/data-crawler.go

```diff
@@ -402,19 +435,92 @@ func (f *folderScanner) scanQueuedLevels(ctx context.Context, folders []cachedFo
 		if err != nil {
 			return nil, err
 		}
+
+		if f.healObjectSelect == 0 {
+			// If we are not scanning, return now.
+			continue
+		}
+
+		// Whatever remains in 'existing' are folders at this level
+		// that existed in the previous run but wasn't found now.
+		//
+		// This may be because of 2 reasons:
+		//
+		// 1) The folder/object was deleted.
+		// 2) We come from another disk and this disk missed the write.
+		//
+		// We therefore perform a heal check.
+		// If that doesn't bring it back we remove the folder and assume it was deleted.
+		// This means that the next run will not look for it.
+		objAPI := newObjectLayerWithoutSafeModeFn()
+		for k := range existing {
+			f.waitForLowActiveIO()
+			bucket, prefix := path2BucketObject(k)
+			if f.dataUsageCrawlDebug {
+				logger.Info(color.Green("folder-scanner:")+" checking disappeared folder: %v/%v", bucket, prefix)
+			}
+			// Try as a folder first.
+			//
+			// A folder on the disk can either be a prefix level or an object.
+			// There is no way to tell, and since it is missing on this disk we cannot cheaply check.
+			// If it is a prefix folder the call below will restore any objects in it.
+			// However, if the path is an object itself this will not be restored.
+			err := objAPI.HealObjects(ctx, bucket, prefix+slashSeparator, madmin.HealOpts{Recursive: true, Remove: healDeleteDangling},
+				func(bucket, object, versionID string) error {
+					_, err := objAPI.HealObject(ctx, bucket, object, versionID, madmin.HealOpts{Recursive: true, Remove: healDeleteDangling})
+					return err
+				})
+			if err != nil {
+				// If path is an object nil is still returned.
+				logger.LogIf(ctx, err)
+				continue
+			}
+
+			// If it still doesn't exist it may be an object.
+			_, err = os.Stat(filepath.Join(f.root, k))
+			if os.IsNotExist(err) {
+				// The 'HealObjects' above will not work if the folder only contains an object.
```

Comment:
The way to do that is you can initiate a HealObjects() by simply calling this on a parent, my suggestion is we can simplify this by just calling one HealObject() in all cases.

HealObjects() is not needed at all - unless for some reason crawler never reaches the leaf i.e the object deep down and at times can have folders is that the case why we have to do this two-step process?

Reply:
If that is the case and we can rely on prefix we can verify two things one is 

- the prefix ends with an object we proceed to initiate HealObjects() regardless
- the prefix doesn't end with '/' then we do path.Dir() and initiate a HealObjects() on the parent instead

In both situations we can simply rely on the fact of HealObjects() instead of HealObject failover it would work all the time and will be consistent.

Reply:
@harshavardhana Calling healobjects on the parent is much more expensive than calling on what we know is missing.

If you make a bucket with 100k objects for instance. If an object is deleted we would check 100k objects instead of just checking the one we know is missing. If we blindly call healobjects on the parent we gain nothing (and it would be completely unfeasible to to).

> `the prefix ends with ...`

We have no idea. We are crawling a folder on disk and we have no idea if what is beneath is a folder or an object..

You are assuming we know the prefix. We don't. We have a path on disk, so it can be anything.

Reply:
That is a logic we can ascertain because you are relying on things like os.Stat() already @klauspost if this code already works we can proceed further by rebasing and trying to see if there are optimizations in the future.

Are there any knowns problems that you faced with this PR?

Reply:
Yes, and it is a bit harder since it is *missing* on this disk. @harshavardhana 

There is a theoretical issue I would like to investigate first. TLDR is that if `HealObjects` fails to recreate anything `HealObject` will always delete the folder on other disks since it sees it as dangling folders and cannot find a `meta.xl`.

But it would only come up if a cluster regains read quorum. I tried to guard against this with the error check on the returned value from `HealObjects`, but maybe if you could confirm that it would return an error when it is unable to get read quorum that would be great.

Reply:
I will check this PR soon in more detail and test @klauspost

Reply:
Thanks!